### ISSUE TITLE
Sagepay response ignoring postVars from Sagepay

### ISSUE THREAD

**Issue by catharsisjelly** [silverstripe/silverstripe-omnipay#159]
Hi all,

I'm debugging the SagePay Server implementation of this. Bear with me on this it's going to require some explaining.

I prepare my purchase which then passes through `PurchaseService::initiate` which then sends me as expected to SagePay server. Once I finish with Sagepay it returns me to the `returnUrl` or as SagePay know it the `NotificationURL` which is `http://mysite.com/paymentendpoint/{Identifier}/complete`. Sagepay send back a fair bit of information in the POST, below is a dump of this info from `PaymentGatewayController` inside `$this->request->postVars()`.

```
array (
  'VPSProtocol' => '3.00',
  'TxType' => 'PAYMENT',
  'VendorTxCode' => '3-13',
  'VPSTxId' => '{92371F6E-E38D-0E3B-D2AB-040338957C7A}',
  'Status' => 'ABORT',
  'StatusDetail' => '2008 : The Transaction timed-out.',
  'AVSCV2' => 'DATA NOT CHECKED',
  'AddressResult' => 'NOTPROVIDED',
  'PostCodeResult' => 'NOTPROVIDED',
  'CV2Result' => 'NOTPROVIDED',
  'GiftAid' => '0',
  '3DSecureStatus' => 'NOTCHECKED',
  'VPSSignature' => '32284FE320F82D3A597253D995E9C30F',
)
```

As far as I can see this information being sent back from Sagepay is not being used and needs to be.   The `PurchaseService::complete` is called which puts together a request to send back to SagePay.

`ServerCompleteAuthorizeRequest` is then invoked via `->send()` which then looks for a key in the parameters called `transactionReference` which is then assumed to be JSON. Did I miss something or is this a legitimate problem?

I'm looking more into it will post more as soon as I can

**Comment by catharsisjelly** [comment]
As I am debugging to get this info into the Gateway I am currently doing this inside `PaymentGatewayController`

```
$serviceResponse = $service->complete([
    'transactionReference' => json_encode($this->request->postVars())
]);
```

It's still not working but this is something to do with the Omnipay section of the code which I have asked SagePay about.

**Comment by bummzack** [comment]
The data coming back from the server is being used… the omnipay library grabs it directly from the superglobals (via Symphony classes).

What exactly is the issue you're facing? Do payments not complete? Did you investigate why you got a `2008 : The Transaction timed-out.` status?

**Comment by catharsisjelly** [comment]
Paymentas don't complete because the code throws an `InvalidRequestException` because the request we send back to Sagepay is missing the `transactionReference` key. This is the Error explained for 2008 in Sagepay - http://www.sagepay.co.uk/support/error-codes/notauthed-transaction-timed-out. I have a support ticket open with them too.

**Comment by bummzack** [comment]
You already seem to get the timeout error when sagepay returns to you, via `http://mysite.com/paymentendpoint/{Identifier}/complete`. So there must be something wrong with your initial request, or with sagepay.

**Comment by catharsisjelly** [comment]
I'll ask them about it. Do you know where the Status that is being passed back from Sagepay is being dealt with as I have not been able to find that either?

**Comment by bummzack** [comment]
I think it would be better to discuss this in chat. Can you contact me via [Gitter](https://gitter.im/silverstripe/silverstripe-omnipay) or on [Slack](https://silverstripe-users.slack.com/)?

**Comment by catharsisjelly** [comment]
Heya,

Yep, I will try to catch you on Gitter sometime today. After sleeping on this and trying this morning I got an okay response back from Sagepay this morning. I think I was getting an ABORT because I was using a non-3d secure card and the account that I am using has 3d Secure set. So my response is now. So now I have that I will remove my hacks and see how things go.

**Comment by catharsisjelly** [comment]
Now that I have a good response back from SagePay I wanted to update this. Below is the response I am getting on an accepted payment

```
array (
  'VPSProtocol' => '3.00',
  'TxType' => 'PAYMENT',
  'VendorTxCode' => '4-1',
  'VPSTxId' => '{EA0DD060-22B6-28FD-0BBD-53842ECEA9D6}',
  'Status' => 'OK',
  'StatusDetail' => '0000 : The Authorisation was Successful.',
  'TxAuthNo' => '14055059',
  'AVSCV2' => 'ALL MATCH',
  'AddressResult' => 'MATCHED',
  'PostCodeResult' => 'MATCHED',
  'CV2Result' => 'MATCHED',
  'GiftAid' => '0',
  '3DSecureStatus' => 'OK',
  'CAVV' => '00000101147993000000000000000000',
  'CardType' => 'VISA',
  'Last4Digits' => '0006',
  'VPSSignature' => 'FB310615452FEB2A9AD0D5DC4F18D063',
  'DeclineCode' => '00',
  'ExpiryDate' => '1017',
  'BankAuthCode' => '999777',
)
```

This is missing a `SecurityKey` which I have asked SagePay about. The code is now throwing an `OmnipayException` because `The transactionReference parameter is required`.

**Comment by catharsisjelly** [comment]
After a lot of debugging and a weekend of not looking at this I'm still looking into why it's not working. I have cloned the code and made the change I pointed out to the `GatewayMessage` table but would like to see it actually working before I submit this with a fix and maybe some documentation additions.

**Comment by catharsisjelly** [comment]
After more debugging, we found that Sagepay is expecting a response to its notification. With a lot of help from @bummzack, we finally got a payment to go through. I will submit a PR with a SagePay extension that should help future Sagepay users

**Comment by bummzack** [comment]
Closed via #163 and released in 2.2.0
